src/asami/projection.cljc

```diff
@@ -203,6 +203,32 @@
           (throw (ex-info (str "Projection variables not found in the selected data: " missing)
                           {:missing missing :data columns})))))))
 
+(s/defn derive-pattern-and-columns :- (s/pair [s/Any] :pattern
+                                              [s/Any] :columns)
+  "Given a pattern (possibly with usage of the :as syntax), return a
+  pair of a pattern without usage of the :as syntax and renamed
+  columns as described by the :as syntax.
+
+  Example:
+
+    (derive-pattern-and-columns '[?x :data :as ?z ?y])
+    ;; => [[?x :data ?y] [?x ?z ?y]]"
+  [pattern]
+  (loop [pattern pattern
+         new-pattern []
+         renamed-columns []]
+    (if (seq pattern)
+      (let [[x y z :as xs] (take 3 pattern)
```

Comment:
This works, but I have 2 questions:
- Why save the triple in `xs`?
- Why use `take`? You can just use `pattern` here

Reply:
Yes, you're right. That's leftover from my first swing at it. I'll clean it up.